rpc/nodedialer: TestConnHealthTryDial failed

rpc/nodedialer.TestConnHealthTryDial [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3575732&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3575732&tab=artifacts#/) on master @ [fbc092b88893bfe4adb5285924614bf8e5132e4f](https://github.com/cockroachdb/cockroach/commits/fbc092b88893bfe4adb5285924614bf8e5132e4f):


```
=== RUN   TestConnHealthTryDial
I211013 03:41:25.182049 303 2@rpc/context.go:1045  [-] 23  dialing n1: 127.0.0.1:39225 (default)
    nodedialer_test.go:189: 
        	Error Trace:	nodedialer_test.go:189
        	Error:      	Not equal: 
        	            	expected: *withstack.withStack(&withstack.withStack{
        	            	    cause: &errutil.leafError{msg:"not yet heartbeated"},
        	            	    stack: &withstack.stack{0x16b6290, 0x45978c, 0x459712, 0x459712, 0x44bae5, 0x482fc1},
        	            	})
        	            	actual  : <nil>(<nil>)
        	Test:       	TestConnHealthTryDial
I211013 03:41:25.184011 41 2@rpc/context.go:688  [-] 24  closing [{n1: 127.0.0.1:39225 (default)} {n0: 127.0.0.1:39225 (default)}]
--- FAIL: TestConnHealthTryDial (0.03s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestConnHealthTryDial PKG=./pkg/rpc/nodedialer TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/server erikgrinaker
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConnHealthTryDial.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Epic CRDB-11087

--- a/pkg/rpc/nodedialer/nodedialer_test.go
+++ b/pkg/rpc/nodedialer/nodedialer_test.go
@@ -184,9 +184,11 @@ func TestConnHealthTryDial(t *testing.T) {
 	defer stopper.Stop(ctx)
 	nd := New(rpcCtx, newSingleNodeResolver(staticNodeID, ln.Addr()))
 
+	// Make sure no connection exists yet, via ConnHealth().
+	require.Equal(t, rpc.ErrNoConnection, nd.ConnHealth(staticNodeID, rpc.DefaultClass))
+
 	// When no connection exists, we expect ConnHealthTryDial to dial the node,
 	// which will return ErrNoHeartbeat at first but eventually succeed.
-	require.Equal(t, rpc.ErrNotHeartbeated, nd.ConnHealthTryDial(staticNodeID, rpc.DefaultClass))
 	require.Eventually(t, func() bool {
 		return nd.ConnHealthTryDial(staticNodeID, rpc.DefaultClass) == nil
 	}, time.Second, 10*time.Millisecond)

